q2_quality_control/decontam.py

```diff
@@ -0,0 +1,163 @@
+    'method': _DECON_METHOD_STR,
+    'freq_concentration_column': _SKIP,
+    'prev_control_column': _SKIP,
+    'prev_control_indicator': _SKIP,
+    'decontam_scores': _SKIP,
+}
+
+
+def _check_inputs(**kwargs):
+    for param, arg in kwargs.items():
+        check_is_valid, explanation = _valid_inputs[param]
+        if not check_is_valid(arg):
+            raise ValueError('Argument to %r was %r, should be %s.'
+                             % (param, arg, explanation))
+
+
+def _check_column_inputs(metadata, method, freq_concentration_column,
+                         prev_control_column, prev_control_indicator):
+    if method == 'prevalence':
+        if prev_control_column not in metadata.columns:
+            raise ValueError('Prevalence column not found, please '
+                             'select from:\n'
+                             + str(', '.join(metadata.columns)))
+        else:
+            if prev_control_indicator not in list(
+                    metadata[prev_control_column]):
+                raise ValueError('No control values found, please select '
+                                 'from:\n' +
+                                 str(', '.join(
+                                     metadata[prev_control_column]
+                                     .unique())))
+            else:
+                print("All inputs are found")
+    elif method == 'frequency':
+        if freq_concentration_column not in metadata.columns:
+            raise ValueError('Frequency column not found, please '
+                             'select from:\n'
+                             + str(', '.join(metadata.columns)))
+        else:
+            print("All inputs are found")
+    else:
+        if ((freq_concentration_column not in metadata.columns) or
+                (prev_control_column not in metadata.columns)):
+            raise ValueError('Column id input error, please '
+                             'select from:\n'
+                             + str(', '.join(metadata.columns)))
+        else:
+            if prev_control_indicator not in list(
+                    metadata[prev_control_column]):
+                raise ValueError('No control values found, please select '
+                                 'from:\n' +
+                                 str(', '.join(
+                                     metadata[prev_control_column]
+                                     .unique())))
+            else:
+                print("All inputs are found")
+
+
+def _decontam_identify_helper(track_fp, method):
+    df = pd.read_csv(track_fp, sep='\t', index_col=0)
+    df.index.name = '#OTU ID'
+    # removes last column containing true/false information from the dataframe
+    df = df.drop(df.columns[(len(df.columns)-1)], axis=1)
+    if method == 'combined':
+        df = df.fillna(0)
+
+    # removes all columns that are completely empty
+    temp_transposed_table = df.transpose()
+    temp_transposed_table = temp_transposed_table.dropna()
```

Comment:
For reasons I don't entirely understand, I'm ending up with something like:
```
                                          freq  prev  p.freq  p.prev   p  contaminant
0001d123420b59585627edf5a1292ae8  6.206727e-05    11     NaN     NaN NaN        False
000a61b8f927a303e169679c636d8f6c  2.452755e-07     1     NaN     NaN NaN        False
000ada501122ff6c83799b3565d1d9e7  3.335465e-05    16     NaN     NaN NaN        False
000b37e67b2e6ff90ef6e3949a402e70  4.938187e-07     1     NaN     NaN NaN        False
000b3d48bd30cf60377e986d8d727c5f  1.249071e-07     1     NaN     NaN NaN        False
...                                        ...   ...     ...     ...  ..          ...
```
which causes `p` to be dropped, breaking the downstream viz in an unusual way. Of course, the more fundamental issue is the `NaN` for `p`, but it's extra mysterious without the column.
```

Reply:
Ah! I discovered it's because the table I started with was filtered to a minimum frequency of 10k features, so it was the same as not passing any controls at all, since they were absent from the table.

I'm pretty sure that's why we're getting NaN. Perhaps the way to address that would be to confirm that the set of control samples is non-empty and error if so. (I'm sure I won't be the only one to run into this accidentally).

Reply:
I am a bit confused about what you mean here. Do you mean that you passed in a table and there were no designated control samples which gave `NaN` when using the `prevalence` and `combined`  decontam methods?

Reply:
Yes, I believe that's what ultimately happened. There were control samples originally, but they were dropped from the table, while existing in the sample metadata file.

Reply:
Ohhh i see, so you are saying that in a table with more than 10K features (columns) then they are being dropped from the table? I think that this could be happening when passing the data from python to R since I used a temp folder and csv files which can only have around ~16K columns in them. Does that sound right from what you are seeing on your side?

Reply:
No its totally fine, I appreciate you taking the time to answer all my questions! I was just confused as to what step the dropping of features was occurring, but I think I got it now. So the issue is that the control samples are present in the metadata file but not in the input ASV table?

Reply:
Yep!